src/app/blog/0rtt-api/page.mdx

```diff
@@ -0,0 +1,394 @@
+2025-05-22T15:10:03.317830Z TRACE 0rtt: 0-RTT possible from our side
+
+```
+
+Now this is weird.
+
+The initial connection is not 0-RTT, which is expected. The two nodes haven't ever talked, so there is no pre shared secret they can use to communicate.
+
+The next two rounds are 0-RTT, as expected, and have roughly the expected roundtrip times.
+
+But then we get a non 0-RTT connection on the 4th round. What's going on?
+
+The issue is that 0-RTT is done using *single-use* tickets. *After* each completed handshake, you get a number of these tickets, by default 2. And as soon as the tickets are used up, the client can not even *attempt* to do a 0-rtt request because it has run out of tickets to use.
+
+In the code so far we close the connection immediately after receiving the user data response. So we don't wait for the full handshake to complete, and also don't get any additional tickets.
+
+```rust
+pingpong_0rtt(connecting, i).await?
+connection.close(0u8.into(), b"");
+```
+
+There are three ways to solve this.
+
+## Waiting for new tickets
+
+The tickets for 0-RTT connections are single use in Quinn, as recommended in the TLS spec: https://datatracker.ietf.org/doc/html/rfc8446#appendix-C.4 . We can get new tickets via a connection that was initiated with 0-RTT, but only if we allow time for the handshake to complete.
+
+There is currently no API to wait for the handshake to complete or to wait for session tickets. But we can just sleep for two times the current round trip time before closing the connection. This is a crude way to allow for enough time for the two `NewSessionTicket`s to be received.
+
+Since we don't want to hold up the user space just because we are waiting for an additional message, we do this in a tokio task.
+
+```rust
+tokio::spawn(async move {
+  tokio::time::sleep(connection.rtt() * 2).await;
+  connection.close(0u8.into(), b"");
+})
+```
+
+In the future we would want to replace the crude sleep with a more precise API that waits exactly until at least one NewSessionTicket message has been received.
+
+<div className="not-prose">
+  <img src="/blog/0rtt-api/0rtt-renew.png" width={1200} height={754} alt="0rtt communication with ticket renewal" />
+</div>
+
+## Sending more tickets
+
+Another possible solution would be to just reconfigure the server to send more session tickets. E.g. you could send 8 instead of 2. The downside of this is that you send additional data, but the upside is that you get more 0-RTT connections per full handshake. Alas, this is currently not configurable without having to modify iroh.
+
+## Reusing tickets
+
+The TLS spec says that you *SHOULD NOT* reuse tickets - you should treat them as single use on the client side. But *SHOULD NOT* just *discourages* this, but does not outright forbid it. And due to its stateless nature the server side will not reject multiple 0-RTT connection attempts using the same ticket.
+
+So in theory you could modify the client to reuse the ticket.
```

Comment:
How long will the server keep around that ticket though?  Doesn't it discard it once used?  I suspect there might be more to it than just the client-side here, so the suggestion might be a bit naive.  (Plus, what is the wisdom of suggesting to folks it ignore a SHOULD NOT?  That's a pretty strong instruction.)

I don't think the article would lose anything if this short section was removed entirely.  Plus I'm a little worried about iroh's reputation when we're suggesting to ignore SHOULDs in specs.  But I know it's hard to resist poking at the loopholes.

Reply:
I don't recommend removing the single use, just mention it as an option. At least from my brief look at the spec and talking with the robots, the server will **not** remove a ticket once used because there was a conscious decision to have less server side state.